mayachemtools/docs/modules/man3/HydrogenBondsDescriptors.3
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "HydrogenBondsDescriptors 1"
.TH HydrogenBondsDescriptors 1 "2018-05-15" "perl v5.22.4" "MayaChemTools"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
HydrogenBondsDescriptors
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
use MolecularDescriptors::HydrogenBondsDescriptors;
.PP
use MolecularDescriptors::HydrogenBondsDescriptors qw(:all);
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBHydrogenBondsDescriptors\fR class provides the following methods:
.PP
new, GenerateDescriptors, GetDescriptorNames, SetHydrogenBondsType,
StringifyHydrogenBondsDescriptors
.PP
\&\fBHydrogenBondsDescriptors\fR is derived from \fBMolecularDescriptors\fR class which in turn
is  derived from \fBObjectProperty\fR base class that provides methods not explicitly defined
in \fBHydrogenBondsDescriptors\fR, \fBMolecularDescriptors\fR or \fBObjectProperty\fR classes using Perl's
\&\s-1AUTOLOAD\s0 functionality. These methods are generated on-the-fly for a specified object property:
.PP
.Vb 3
\&    Set<PropertyName>(<PropertyValue>);
\&    $PropertyValue = Get<PropertyName>();
\&    Delete<PropertyName>();
.Ve
.PP
The current release of MayaChemTools supports identification of two types of hydrogen bond
donor and acceptor atoms with these names:
.PP
.Vb 2
\&    HBondsType1 or HydrogenBondsType1
\&    HBondsType2 or HydrogenBondsType2
.Ve
.PP
The names of these hydrogen bond types are rather arbitrary. However, their definitions have
specific meaning and are as follows:
.PP
.Vb 1
\&    HydrogenBondsType1 [ Ref 60\-61, Ref 65\-66 ]:
\&
\&        Donor: NH, NH2, OH \- Any N and O with available H
\&        Acceptor: N[!H], O \- Any N without available H and any O
\&
\&    HydrogenBondsType2 [ Ref 91 ]:
\&
\&        Donor: NH, NH2, OH \- N and O with available H
\&        Acceptor: N, O \- And N and O
.Ve
.PP
By default, \fIHydrogenBondsType2\fR is used to calculate number hydrogen bond donor
and acceptor atoms. This corresponds to \fBRuleOf5\fR definition of hydrogen bond donors
and acceptors.
.SS "\s-1METHODS\s0"
.IX Subsection "METHODS"
.IP "\fBnew\fR" 4
.IX Item "new"
.Vb 2
\&    $HydrogenBondsDescriptors = new MolecularDescriptors::
\&                                HydrogenBondsDescriptors(%NamesAndValues);
.Ve
.Sp
Using specified \fIHydrogenBondsDescriptors\fR property names and values hash, \fBnew\fR
method creates a new object and returns a reference to newly created \fBHydrogenBondsDescriptors\fR
object. By default, the following properties are initialized:
.Sp
.Vb 5
\&    Molecule = \*(Aq\*(Aq
\&    Type = \*(AqHydrogenBonds\*(Aq
\&    HydrogenBondsType = \*(AqHBondsType2\*(Aq
\&    @DescriptorNames = (\*(AqHydrogenBondDonors\*(Aq, \*(AqHydrogenBondAcceptors\*(Aq)
\&    @DescriptorValues = (\*(AqNone\*(Aq, \*(AqNone\*(Aq)
.Ve
.Sp
Examples:
.Sp
.Vb 2
\&    $HydrogenBondsDescriptors = new MolecularDescriptors::
\&                                HydrogenBondsDescriptors();
\&
\&    $HydrogenBondsDescriptors = new MolecularDescriptors::
\&                                HydrogenBondsDescriptors(
\&                                \*(AqHydrogenBondsType\*(Aq => \*(AqHBondsType2\*(Aq);
\&
\&    $HydrogenBondsDescriptors\->SetMolecule($Molecule);
\&    $HydrogenBondsDescriptors\->GenerateDescriptors();
\&    print "HydrogenBondsDescriptors: $HydrogenBondsDescriptors\en";
.Ve
.IP "\fBGenerateDescriptors\fR" 4
.IX Item "GenerateDescriptors"
.Vb 1
\&    $HydrogenBondsDescriptors\->GenerateDescriptors();
.Ve
.Sp
Calculates number of hydrogen bond donors and acceptors a molecule and returns
\&\fIHydrogenBondsDescriptors\fR.
.IP "\fBGetDescriptorNames\fR" 4
.IX Item "GetDescriptorNames"
.Vb 3
\&    @DescriptorNames = $HydrogenBondsDescriptors\->GetDescriptorNames();
\&    @DescriptorNames = MolecularDescriptors::HydrogenBondsDescriptors::
\&                       GetDescriptorNames();
.Ve
.Sp
Returns all available descriptor names as an array.
.IP "\fBSetHydrogenBondsType\fR" 4
.IX Item "SetHydrogenBondsType"
.Vb 1
\&    $HydrogenBondsDescriptors\->SetHydrogenBondsType($HBondsType);
.Ve
.Sp
Sets value of hydrogen bonds type to use during calculation of descriptors and returns
\&\fIHydrogenBondsDescriptors\fR. Possible values: \fIHBondsType1, HydrogenBondsType1,
HBondsType2, HydrogenBondsType2\fR.
.IP "\fBStringifyHydrogenBondsDescriptors\fR" 4
.IX Item "StringifyHydrogenBondsDescriptors"
.Vb 2
\&    $String = $HydrogenBondsDescriptors\->
\&                              StringifyHydrogenBondsDescriptors();
.Ve
.Sp
Returns a string containing information about \fIHydrogenBondsDescriptors\fR object.
.SH "AUTHOR"
.IX Header "AUTHOR"
Manish Sud <msud@san.rr.com>
.SH "SEE ALSO"
.IX Header "SEE ALSO"
MolecularDescriptors.pm, MolecularDescriptorsGenerator.pm
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (C) 2018 Manish Sud. All rights reserved.
.PP
This file is part of MayaChemTools.
.PP
MayaChemTools is free software; you can redistribute it and/or modify it under
the terms of the \s-1GNU\s0 Lesser General Public License as published by the Free
Software Foundation; either version 3 of the License, or (at your option)
any later version.


scripts/mayachemtools/lib/MolecularDescriptors/HydrogenBondsDescriptors.pm
package MolecularDescriptors::HydrogenBondsDescriptors;
#
# File: HydrogenBondsDescriptors.pm
# Author: Manish Sud <msud@san.rr.com>
#
# Copyright (C) 2018 Manish Sud. All rights reserved.
#
# This file is part of MayaChemTools.
#
# MayaChemTools is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation; either version 3 of the License, or (at your option) any
# later version.
#
# MayaChemTools is distributed in the hope that it will be useful, but without
# any warranty; without even the implied warranty of merchantability of fitness
# for a particular purpose.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with MayaChemTools; if not, see <http://www.gnu.org/licenses/> or
# write to the Free Software Foundation Inc., 59 Temple Place, Suite 330,
# Boston, MA, 02111-1307, USA.
#

use strict;
use Carp;
use Exporter;
use Scalar::Util ();
use TextUtil ();
use Atom;
use Molecule;
use MolecularDescriptors::MolecularDescriptors;

use vars qw(@ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);

@ISA = qw(MolecularDescriptors::MolecularDescriptors Exporter);
@EXPORT = qw();
@EXPORT_OK = qw(GetDescriptorNames);

%EXPORT_TAGS = (all  => [@EXPORT, @EXPORT_OK]);

# Setup class variables...
my($ClassName, @DescriptorNames);
_InitializeClass();

# Overload Perl functions...
use overload '""' => 'StringifyHydrogenBondsDescriptors';

# Class constructor...
sub new {
  my($Class, %NamesAndValues) = @_;

  # Initialize object...
  my $This = $Class->SUPER::new();
  bless $This, ref($Class) || $Class;
  $This->_InitializeHydrogenBondsDescriptors();

  $This->_InitializeHydrogenBondsDescriptorsProperties(%NamesAndValues);

  return $This;
}

# Initialize class ...
sub _InitializeClass {
  #Class name...
  $ClassName = __PACKAGE__;

  # Descriptor names...
  @DescriptorNames = ('HydrogenBondDonors', 'HydrogenBondAcceptors');

}

# Get descriptor names as an array.
#
# This functionality can be either invoked as a class function or an
# object method.
#
sub GetDescriptorNames {
  return @DescriptorNames;
}

# Initialize object data...
#
sub _InitializeHydrogenBondsDescriptors {
  my($This) = @_;

  # Type of MolecularDescriptor...
  $This->{Type} = 'HydrogenBonds';

  # The currrent release of MayaChemTools supports identification of two types of
  # hydrogen bond donor and acceptor atoms with these names:
  #
  # HBondsType1 or HydrogenBondsType1
  # HBondsType2 or HydrogenBondsType2
  #
  # The names of these hydrogen bond types are rather arbirary. However, their
  # definitions have specific meaning and are as follows:
  #
  # HydrogenBondsType1 [ Ref 60-61, Ref 65-66 ]:
  #   . Donor: NH, NH2, OH - Any N and O with available H
  #   . Acceptor: N[!H], O - Any N without available H and any O
  #
  # HydrogenBondsType2 [ Ref 91 ]:
  #   . Donor: NH, NH2, OH - N and O with availabe H
  #   . Acceptor: N, O - Add N and O
  #
  # Note:
  #   . HydrogenBondsType2 definition corresponds to Rule of 5.
  #
  $This->{HydrogenBondsType} = 'HBondsType2';

  # Intialize descriptor names and values...
  $This->_InitializeDescriptorNamesAndValues(@DescriptorNames);

  return $This;
}

# Initialize object properties...
#
sub _InitializeHydrogenBondsDescriptorsProperties {
  my($This, %NamesAndValues) = @_;

  my($Name, $Value, $MethodName);
  while (($Name, $Value) = each  %NamesAndValues) {
    $MethodName = "Set${Name}";
    $This->$MethodName($Value);
  }

  return $This;
}

# Set hydrogen bonds type...
#
sub SetHydrogenBondsType {
  my($This, $HydrogenBondsType) = @_;

  if ($HydrogenBondsType !~ /^(HBondsType1|HBondsType2|HydrogenBondsType1|HydrogenBondsType2)$/i) {
    croak "Error: ${ClassName}->SetHydrogenBondsType: Specified hydrogen bonds type, $HydrogenBondsType, is not supported. Valid values: HBondsType1, HBondsType2, HydrogenBondsType1, HydrogenBondsType2 ...\n ";
  }

  $This->{HydrogenBondsType} = $HydrogenBondsType;

  return $This;
}

# Calculate number of hydrogen bond donors and acceptors in a molecule...
#
sub GenerateDescriptors {
  my($This) = @_;

  # Initialize descriptor values...
  $This->_InitializeDescriptorValues();

  # Check availability of molecule...
  if (!$This->{Molecule}) {
    carp "Warning: ${ClassName}->GenerateDescriptors: $This->{Type} molecular descriptors generation didn't succeed: Molecule data is not available: Molecule object hasn't been set...";
    return undef;
  }

  # Calculate descriptor values...
  if (!$This->_CalculateDescriptorValues()) {
    carp "Warning: ${ClassName}->CalculateDescriptorValues: $This->{Type} molecular descriptors generation didn't succeed: Couldn't calculate number of hydrogen bond donor and accepror values...";
    return undef;
  }

  # Set final descriptor values...
  $This->_SetFinalDescriptorValues();

}

# Calculate number of hydrogen bond donors and acceptors...
#
sub _CalculateDescriptorValues {
  my($This) = @_;
  my($HydrogenBondDonors, $HydrogenBondAcceptors, $Atom);

  $HydrogenBondDonors = 0;
  $HydrogenBondAcceptors = 0;

  for $Atom ($This->{Molecule}->GetAtoms()) {
    if ($Atom->IsHydrogenBondDonor($This->{HydrogenBondsType})) {
      $HydrogenBondDonors++;
    }
    if ($Atom->IsHydrogenBondAcceptor($This->{HydrogenBondsType})) {
      $HydrogenBondAcceptors++;
    }
  }

  # Track the calculated values...
  $This->{HydrogenBondDonors} = $HydrogenBondDonors;
  $This->{HydrogenBondAcceptors} = $HydrogenBondAcceptors;

  return $This;
}

# Setup final descriptor values...
#
sub _SetFinalDescriptorValues {
  my($This) = @_;

  $This->{DescriptorsGenerated} = 1;

  $This->SetDescriptorValues($This->{HydrogenBondDonors}, $This->{HydrogenBondAcceptors});

  return $This;
}

# Return a string containg data for HydrogenBondsDescriptors object...
#
sub StringifyHydrogenBondsDescriptors {
  my($This) = @_;
  my($HydrogenBondsDescriptorsString);

  $HydrogenBondsDescriptorsString = "MolecularDescriptorType: $This->{Type}; HydrogenBondsType: $This->{HydrogenBondsType}; " . $This->_StringifyDescriptorNamesAndValues();

  return $HydrogenBondsDescriptorsString;
}

# Is it a HydrogenBondsDescriptors object?
sub _IsHydrogenBondsDescriptors {
  my($Object) = @_;

  return (Scalar::Util::blessed($Object) && $Object->isa($ClassName)) ? 1 : 0;
}

1;

__END__

=head1 NAME

HydrogenBondsDescriptors

=head1 SYNOPSIS

use MolecularDescriptors::HydrogenBondsDescriptors;

use MolecularDescriptors::HydrogenBondsDescriptors qw(:all);

=head1 DESCRIPTION

B<HydrogenBondsDescriptors> class provides the following methods:

new, GenerateDescriptors, GetDescriptorNames, SetHydrogenBondsType,
StringifyHydrogenBondsDescriptors

B<HydrogenBondsDescriptors> is derived from B<MolecularDescriptors> class which in turn
is  derived from B<ObjectProperty> base class that provides methods not explicitly defined
in B<HydrogenBondsDescriptors>, B<MolecularDescriptors> or B<ObjectProperty> classes using Perl's
AUTOLOAD functionality. These methods are generated on-the-fly for a specified object property:

    Set<PropertyName>(<PropertyValue>);
    $PropertyValue = Get<PropertyName>();
    Delete<PropertyName>();

The current release of MayaChemTools supports identification of two types of hydrogen bond
donor and acceptor atoms with these names:

    HBondsType1 or HydrogenBondsType1
    HBondsType2 or HydrogenBondsType2

The names of these hydrogen bond types are rather arbitrary. However, their definitions have
specific meaning and are as follows:

    HydrogenBondsType1 [ Ref 60-61, Ref 65-66 ]:

        Donor: NH, NH2, OH - Any N and O with available H
        Acceptor: N[!H], O - Any N without available H and any O

    HydrogenBondsType2 [ Ref 91 ]:

        Donor: NH, NH2, OH - N and O with available H
        Acceptor: N, O - And N and O

By default, I<HydrogenBondsType2> is used to calculate number hydrogen bond donor
and acceptor atoms. This corresponds to B<RuleOf5> definition of hydrogen bond donors
and acceptors.

=head2 METHODS

=over 4

=item B<new>

    $HydrogenBondsDescriptors = new MolecularDescriptors::
                                HydrogenBondsDescriptors(%NamesAndValues);

Using specified I<HydrogenBondsDescriptors> property names and values hash, B<new>
method creates a new object and returns a reference to newly created B<HydrogenBondsDescriptors>
object. By default, the following properties are initialized:

    Molecule = ''
    Type = 'HydrogenBonds'
    HydrogenBondsType = 'HBondsType2'
    @DescriptorNames = ('HydrogenBondDonors', 'HydrogenBondAcceptors')
    @DescriptorValues = ('None', 'None')

Examples:

    $HydrogenBondsDescriptors = new MolecularDescriptors::
                                HydrogenBondsDescriptors();

    $HydrogenBondsDescriptors = new MolecularDescriptors::
                                HydrogenBondsDescriptors(
                                'HydrogenBondsType' => 'HBondsType2');

    $HydrogenBondsDescriptors->SetMolecule($Molecule);
    $HydrogenBondsDescriptors->GenerateDescriptors();
    print "HydrogenBondsDescriptors: $HydrogenBondsDescriptors\n";

=item B<GenerateDescriptors>

    $HydrogenBondsDescriptors->GenerateDescriptors();

Calculates number of hydrogen bond donors and acceptors a molecule and returns
I<HydrogenBondsDescriptors>.

=item B<GetDescriptorNames>

    @DescriptorNames = $HydrogenBondsDescriptors->GetDescriptorNames();
    @DescriptorNames = MolecularDescriptors::HydrogenBondsDescriptors::
                       GetDescriptorNames();

Returns all available descriptor names as an array.

=item B<SetHydrogenBondsType>

    $HydrogenBondsDescriptors->SetHydrogenBondsType($HBondsType);

Sets value of hydrogen bonds type to use during calculation of descriptors and returns
I<HydrogenBondsDescriptors>. Possible values: I<HBondsType1, HydrogenBondsType1,
HBondsType2, HydrogenBondsType2>.

=item B<StringifyHydrogenBondsDescriptors>

    $String = $HydrogenBondsDescriptors->
                              StringifyHydrogenBondsDescriptors();

Returns a string containing information about I<HydrogenBondsDescriptors> object.

=back

=head1 AUTHOR

Manish Sud <msud@san.rr.com>

=head1 SEE ALSO

MolecularDescriptors.pm, MolecularDescriptorsGenerator.pm

=head1 COPYRIGHT

Copyright (C) 2018 Manish Sud. All rights reserved.

This file is part of MayaChemTools.

MayaChemTools is free software; you can redistribute it and/or modify it under
the terms of the GNU Lesser General Public License as published by the Free
Software Foundation; either version 3 of the License, or (at your option)
any later version.

=cut
